### ISSUE TITLE
View dependency check bypassed on db reload

### ISSUE THREAD

**Issue by nicku33** [duckdb/duckdb#12239]
### What happens?

With `set enable_view_dependencies=true`, and opening an on disk duckdb file,
after creating a dependency from view `v1` against table `t1`, an attempt to rename  the table fails as expected.
However reloading the database seems to allow one to bypass the check.

I ran this against a build of HEAD as of this moment.
```
> git log -1
commit 1323e95677e571cfacd56e3dd1fe6d3220a76512 (HEAD -> main, origin/main)
Merge: b4c377e6bb 7dc2f1863c
Author: Mark <mark.raasveldt@gmail.com>
Date:   Fri May 24 22:11:50 2024 +0200
```

### To Reproduce

```
> ./duckdb-1323e95677
v0.10.4-dev37 1323e95677
D .open x.ddb
D set enable_view_dependencies=true;
D create table t1 as select 1 as a;
D create view v1 as select * from t1;
D alter table t1 rename to t1_old;

Dependency Error: Cannot alter entry "t1" because there are entries that depend on it.
^^ expected in this mode

> ./duckdb-1323e95677
v0.10.4-dev37 1323e95677
D set enable_view_dependencies=true;   -- set before and after open just in case
D .open x.ddb
D set enable_view_dependencies=true;   
D alter table t1 rename to t1_old;

^^ no error ?

D select * from v1;
Catalog Error: Table with name t1 does not exist!
Did you mean "t1"?
```

### OS:

aarch64 Mac OS X 12.3.1

### DuckDB Version:

v0.10.4-dev37 1323e95677

### DuckDB Client:

CLI

### Full Name:

Nicholas Ursa

### Affiliation:

Motherduck

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a source build

### Did you include all relevant data sets for reproducing the issue?

Not applicable - the reproduction does not require a data set

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by nicku33** [comment]
cc @Tishj @bleskes

**Comment by carlopi** [comment]
I think this ties to the storage_compatibility_version, this appears to work(around) the problem:
```
carlo@ScroogeMcDuck duckdb % ./build/release/duckdb test.db
v0.10.4-dev34 b4c377e6bb
Enter ".help" for usage hints.
D SET storage_compatibility_version='latest';
D SET enable_view_dependencies=true;
D create table t2 as select 1 as a;
D create view v2 as select * from t2;
D alter table t2 rename to t2_old;
Dependency Error: Cannot alter entry "t2" because there are entries that depend on it.
D 
carlo@ScroogeMcDuck duckdb % ./build/release/duckdb test.db
v0.10.4-dev34 b4c377e6bb
Enter ".help" for usage hints.
D alter table t2 rename to t2_old;
Dependency Error: Cannot alter entry "t2" because there are entries that depend on it.
D 
```
Unsure if that's enough for you to move forward. I will tag as reproduced, someone else will take a more proper look.

**Comment by Tishj** [comment]
Carlo is entirely correct

**Comment by Mytherin** [comment]
As mentioned this is tied to the `storage_compatibility_version` and is intentional. The reason for this change is that dependencies used to be re-constructed on start-up by re-binding, but binding on start-up produced a number of problems particularly around `ATTACH` and WAL replays. In order to solve those problems we removed the binding on start-up in https://github.com/duckdb/duckdb/pull/12076. However, as a result, dependencies are no longer reconstructed on start-up and would be lost on restart.

We then added support for serializing the dependencies so they would be preserved. However, since dependencies are very common, and older versions of DuckDB don't know how to deal with serialized dependencies, this change would break forwards compatibility and `v0.10.2` and below would not be able to read files created by `v0.10.3` and above. In order to prevent breaking the forwards compatibility in, we introduced the `storage_compatibility_version` (https://github.com/duckdb/duckdb/pull/12110).

When `storage_compatibility_version` is set to `v0.10.2`, we don't serialize dependencies so that older versions of DuckDB can read the database file. When `storage_compatibility_version` is set to `v0.10.3` or `latest`, dependencies are serialized.

The default version of `storage_compatibility_version` is currently `v0.10.2`, but we will likely increment this default in the future. This all still needs to be documented in a better manner as well.
